group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroupMember.java

```diff
@@ -0,0 +1,499 @@
+            this.metadata = metadata;
+            this.assignment = assignment;
+        }
+
+        public String memberId() {
+            return memberId;
+        }
+
+        public Optional<String> getGroupInstanceId() {
+            return groupInstanceId;
+        }
+
+        public String clientId() {
+            return clientId;
+        }
+
+        public String clientHost() {
+            return clientHost;
+        }
+
+        public byte[] metadata() {
+            return metadata;
+        }
+
+        public byte[] assignment() {
+            return assignment;
+        }
+        
+    }
+
+    /**
+     * The member id.
+     */
+    private final String memberId;
+
+    /**
+     * The group instance id.
+     */
+    private final Optional<String> groupInstanceId;
+
+    /**
+     * The client id.
+     */
+    private final String clientId;
+
+    /**
+     * The client host.
+     */
+    private final String clientHost;
+
+    /**
+     * The rebalance timeout in milliseconds.
+     */
+    private final int rebalanceTimeoutMs;
+
+    /**
+     * The session timeout in milliseconds.
+     */
+    private final int sessionTimeoutMs;
+
+    /**
+     * The protocol type.
+     */
+    private final String protocolType;
+
+    /**
+     * The list of supported protocols.
+     */
+    private final List<Protocol> supportedProtocols;
+
+    /**
+     * The assignment stored by the client assignor.
+     */
+    private final byte[] assignment;
+
+    /**
+     * The callback that is invoked once this member joins the group.
+     */
+    private CompletableFuture<JoinGroupResponseData> awaitingJoinCallback = null;
+
+    /**
+     * The callback that is invoked once this member completes the sync group phase.
+     */
+    private CompletableFuture<SyncGroupResponseData> awaitingSyncCallback = null;
+
+    /**
+     * Indicates whether the member is a new member of the group.
+     */
+    private boolean isNew = false;
+
+    /**
+     * This variable is used to track heartbeat completion through the delayed
+     * heartbeat purgatory. When scheduling a new heartbeat expiration, we set
+     * this value to `false`. Upon receiving the heartbeat (or any other event
+     * indicating the liveness of the client), we set it to `true` so that the
+     * delayed heartbeat can be completed.
+     */
+    private boolean heartbeatSatisfied = false;
+
+
+    public GenericGroupMember(
+        String memberId,
+        Optional<String> groupInstanceId,
+        String clientId,
+        String clientHost,
+        int rebalanceTimeoutMs,
+        int sessionTimeoutMs,
+        String protocolType,
+        List<Protocol> supportedProtocols
+    ) {
+        this(
+            memberId,
+            groupInstanceId,
+            clientId,
+            clientHost,
+            rebalanceTimeoutMs,
+            sessionTimeoutMs,
+            protocolType,
+            supportedProtocols,
+            new byte[0]
```

Comment:
nit: Should we define a constant for this?

Reply:
thanks for the suggestion!